### PR TITLE
[Issue #5] enable user provided schema when reading exasol table

### PR DESCRIPTION
**Pull request by 3cham**

### CODE REVIEW

**Comment by morazow** [comment]
Good morning @3cham,

Thanks for PR! Please keep up the good work!

Overall, everything looks good. However, I have some suggestions. I am going to list them from easier (easy to explain :)) to harder points.

- [ExasolRelation] Remove using `configSchema` in `buildScan()` [here](https://github.com/EXASOL/spark-exasol-connector/pull/19/files#diff-5ff4f1d4461d0e6a1183f6e47dfbc3b3R51). The user schema will be again automatically selected [here](https://github.com/EXASOL/spark-exasol-connector/pull/19/files#diff-5ff4f1d4461d0e6a1183f6e47dfbc3b3R60).

- [ExasolRelation] Keep `querySchema` function as before. But please rename it to `inferSchema`, that describes it clearly.

- [ExasolRelation] Take `configSchema` parameter as option, e.g `configSchema: Option[StructType]`

- [ExasolRelation] Change overrided schema function to get `configSchema` if available and only otherwise inferSchema. Usually, in Scala there are many ways to do this with options. However, let us stick with fold. So in the schema function should be for example:
```scala
// This will use inferSchema if configSchema is None
 override def schema: StructType = configSchema.fold(inferSchema){ userSchema =>
    logger.info(s"Using provided schema $userSchema")
    userSchema 
  }  
```
You can read about it here, http://blog.originate.com/blog/2014/06/15/idiomatic-scala-your-options-do-not-match/, https://kwangyulseo.com/2014/05/21/scala-option-fold-vs-option-mapgetorelse/

- [DefaultSource] Here, let's put query string check and manager creating part into its own private function. E.g,
```scala
private def fromParameters(params: Map...): (String, ExasolManager) = {
    val queryString = parameters.get("query") match ..
    ...
    val manager = ExasolConnectionManager(config)
   (queryString, manager)
}
```
And both `createRelation` function call this, and return respective ExasolRelation.
```scala
 override def createRelation(...): BaseRelation = { 
    val (queryString, manager) = fromParameters(parameters)
    new ExasolRelation(sqlContext, queryString, None, manager)
  }

  override def createRelation(
    ....,                                                       
    schema: StructType                
  ): BaseRelation = {
    val (queryString, manager) = fromParameters(parameters)
    new ExasolRelation(sqlContext, queryString, Option(schema), manager)
  }     
```

Hope these are clear. Please let me know if you have questions!

File: src/main/scala/com/exasol/spark/DefaultSource.scala [hunk 1]
```diff
@@ -1,20 +1,32 @@
 package com.exasol.spark
 
+import java.util.InvalidPropertiesFormatException
```
Comment (morazow) [inline src/main/scala/com/exasol/spark/DefaultSource.scala]: This exception is not used anywhere

File: src/main/scala/com/exasol/spark/ExasolRelation.scala [hunk 2]
```diff
@@ -27,19 +31,24 @@ class ExasolRelation(context: SQLContext, queryString: String, manager: ExasolCo
...
+        Types.createSparkStructType(metadata)
+      }
     }
   }
 
   override def schema: StructType = querySchema
 
   override def buildScan(): RDD[Row] =
-    buildScan(Array.empty, Array.empty)
+    buildScan(configSchema.map(_.name).toArray, Array.empty)
```
[the remark this answers is not in the document]
Comment (3cham) [reply src/main/scala/com/exasol/spark/ExasolRelation.scala]: Hi @morazow, this is where I spent a lot of time to think about. In the ExasolRDD implementation, we use `Converter.resultSetToRows(resultSet, querySchema)`, in this case the jdbc connection still have to fetch all columns including the ones not provided in the configSchema. If the idea is user must provide the full schema that is returns from the queryString, then it is fine to let it like that. However, we could enable restricting jdbc to fetch only the required columns. Do you think it is a good idea?

[the remark this answers is not in the document]
Comment (morazow) [reply src/main/scala/com/exasol/spark/ExasolRelation.scala]: Hey @3cham,

Yes, we only fetch required columns for both cases when schema is provided and is inferred.

I comment below, for better explaining.

[the remark this answers is not in the document]
Comment (morazow) [reply src/main/scala/com/exasol/spark/ExasolRelation.scala]: Ahh I could not comment on un-changed lines :)

Below, on lines 51 / 60 you will see,

```scala
Types.selectColumns(requiredColumns, schema)
```

Here, schema will be either user provided `configSchema` or inferred `inferSchema`. Then if we have required columns we select only those from that `schema`.

[the remark this answers is not in the document]
Comment (3cham) [reply src/main/scala/com/exasol/spark/ExasolRelation.scala]: Ah, yes, I oversaw this line. Thank you for pointing that out :)

**Comment by 3cham** [comment]
Hi @morazow, please check the following up commits. If you don't mind, allow me to change the author info in the last PR ;)

**Comment by morazow** [comment]
Hey @3cham ,
PR looks good! Going to merge soon!

Yes, sure. I was also confused to seeing you in [contributors list](https://github.com/EXASOL/spark-exasol-connector/graphs/contributors). Just for your information, I am also going to add authors list from my fork, https://github.com/morazow/spark-exasol-connector/blob/task/authors/AUTHORS.md.

Best

**Comment by 3cham** [comment]
@morazow: many thanks, I used my lap from work with a different email config. That's why I need to change the info from the last PR.

**Pull request merged**